torch/_dynamo/eval_frame.py

```diff
@@ -262,19 +264,125 @@ def innermost_fn(fn):
+    if hasattr(config_cache, "saved_config"):
+        return
+    else:
+        for c in ["saved_config", "saved_config_hash", "cached_recompile_hash"]:
+            setattr(config_cache, c, None)
+
+
 @contextlib.contextmanager
-def enable_dynamic(enable: Optional[bool] = None, export: bool = False):
-    if enable is None:
+def restore_compiled_dynamo_config(
+    first_ctx: bool,
+    saved_config: Dict[str, Any],
+    saved_config_hash: str,
+):
+    _maybe_init_guarded_config_cache()
+    # Set exactly once from top-level compile
+    is_top_level = False
+    try:
+        if first_ctx and config_cache.saved_config is None:
+            assert config_cache.saved_config_hash is None
+            is_top_level = True
+            config_cache.saved_config = saved_config
+            config_cache.saved_config_hash = saved_config_hash
+            log.debug(
+                "Set top-level compile config hash: %s", config_cache.saved_config_hash
+            )
+        elif config.verbose:
+            log.debug("Ignoring inner dynamo compile config and hash")  # TODO: remove?
         yield
-    elif enable:
-        # Assume everything is dynamic by default
-        with config.patch(assume_static_by_default=False):
-            yield
+    finally:
+        if is_top_level:
+            config_cache.saved_config_hash = config_cache.saved_config = None
+
+
+def get_config_and_hash(dynamic=None):
+    _maybe_init_guarded_config_cache()
+    # get current value of dynamo configs
+    dynamo_config = config.get_config_copy()
+    for k in dynamo_guarded_config_ignorelist:
+        dynamo_config.pop(k)
+
+    # Set appropriate dynamo config flags
+    if dynamic is None:
+        pass
+    elif dynamic:
+        dynamo_config.update({"assume_static_by_default": False})
     else:
-        with config.patch(
-            automatic_dynamic_shapes=False, assume_static_by_default=True
-        ):
-            yield
+        dynamo_config.update(
+            {"automatic_dynamic_shapes": False, "assume_static_by_default": True}
+        )
+
+    serialized_config = dict(dynamo_config)
+
+    # Only string-serialize for the hash
+    for key in dynamo_guarded_config_string_serialization_list:
+        serialized_config.update({key: str(dynamo_config[key])})
+
+    # Make serialization more deterministic by sorting
+    dynamo_config_bytes = pickle.dumps(dict(sorted(serialized_config.items())))
+
+    sha256_hash = hashlib.sha256(dynamo_config_bytes).hexdigest()
+    return dynamo_config, sha256_hash
+
+
+# Used to test if the config needs to be restored by checking the current hash.
+# We cache the latest requested, and only recompute it when config is possibly
+# out of sync (i.e. is dirty).
+def get_cached_recompile_hash():
+    _maybe_init_guarded_config_cache()
+    if config._is_dirty or config_cache.cached_recompile_hash is None:
+        _, config_cache.cached_recompile_hash = get_config_and_hash(dynamic=None)
+        config._is_dirty = False
```

Comment:
Naw man, you can't do this. The config is global, but `config_cache` is TLS. This doesn't make sense. When the config object changes, you need to update ALL of the threads, not just the one that happens to bang on it first.

The most logical fix is to store the cached hash in a global, probably on the config itself. Then you can get rid of dirty too.

Reply:
Yeah, maybe I'll get rid of this optimization and move it into initiatives associated with https://github.com/pytorch/pytorch/issues/111150

Reply:
resolved in https://github.com/pytorch/pytorch/pull/111299